### DISCUSSION TITLE
clean up old virtual builders

### DISCUSSION THREAD

**Discussion by hmoffatt** [buildbot/buildbot#6596] [Q&A]
I have some schedulers that trigger on any branch, so I'd like to use the virtual_builder_name property to split these out in the UI. But will the virtual builders persist forever? Do they ever get hidden if they haven't been used for some time?

**Comment by p12tic** [comment]
If builder is no longer present in the configuration, it should be hidden in the web UI by default.

**Reply by hmoffatt** [reply]
Virtual builders aren't in the configuration though, but created by builds with the right properties set. Is there any mechanism by which they are removed or hidden from the UI when they are no longer in use?

**Reply by hmoffatt** [reply]
The www code explicitly considers any builder tagged `_virtual_` as active too.

**Comment by dgow** [comment]
Hey @hmoffatt do you have found a solution for this? Same here, we are building every pullrequest on a virtual builder. At the moment I'm deleting the builder from database like this:

```python
    # service update method (every some minutes)
    @poll.method
    def updateTick(self):
        conn: BuildersConnectorComponent = self.master.db.builders
        builders = yield conn.getBuilders()

        g = Github(self.token)
        repo = g.get_repo(self.repo)
        prs = repo.get_pulls(state="open")
        # compare open pull requests with running builders
        # remove builder if  corresponding pull request is closed.

...

    def removeBuilder(self, builderId):
        buildersConn: BuildersConnectorComponent = self.master.db.builders

        def thd(conn):
            tbl = buildersConn.db.model.builders
            conn.execute(tbl.delete().where(tbl.c.id == builderId))

        return buildersConn.db.pool.do_with_transaction(thd)
```

it works but feels a bit "dirty" 

any ideas?

**Reply by hmoffatt** [reply]
> Hey @hmoffatt do you have found a solution for this?

No I didn't, and I eventually stopped using virtual builders anyway. 

In my case I would be happy to have them disappear from the UI if they hadn't been used in some time, like 7 days for example.

**Comment by gzzi** [comment]
We have exactly the same issue. Lot of virtual builder and having a way to not show them in the UI OR removing them from the DB would be nice. Maybe the existing Janitor could be used for that.

**Reply by sa2ajj** [reply]
that's pretty much the case: we have plenty of no more virtual builders shown in the UI

**Reply by sa2ajj** [reply]
(i know it does not help much :smile:)
